### ISSUE TITLE
Is it possible to compose existing operators to create new operators

### ISSUE THREAD

**Issue by taoari** [apache/incubator-mxnet#4273]
For example, in TensorFlow, InstanceNorm and BatchNorm can be created as:

```
def instance_norm(x):
    epsilon = 1e-9
    mean, var = tf.nn.moments(x, [1, 2], keep_dims=True)
    return tf.div(tf.sub(x, mean), tf.sqrt(tf.add(var, epsilon)))

def batch_norm(x, size, training, decay=0.999):
    beta = tf.Variable(tf.zeros([size]), name='beta')
    scale = tf.Variable(tf.ones([size]), name='scale')
    pop_mean = tf.Variable(tf.zeros([size]))
    pop_var = tf.Variable(tf.ones([size]))
    epsilon = 1e-3

    batch_mean, batch_var = tf.nn.moments(x, [0, 1, 2])
    train_mean = tf.assign(pop_mean, pop_mean * decay + batch_mean * (1 - decay))
    train_var = tf.assign(pop_var, pop_var * decay + batch_var * (1 - decay))

    def batch_statistics():
        with tf.control_dependencies([train_mean, train_var]):
            return tf.nn.batch_normalization(x, batch_mean, batch_var, beta, scale, epsilon, name='batch_norm')

    def population_statistics():
        return tf.nn.batch_normalization(x, pop_mean, pop_var, beta, scale, epsilon, name='batch_norm')

    return tf.cond(training, batch_statistics, population_statistics)
```

I am just wondering if this is possible in MXNet. If possible, is there any example for illustrating this? 

The good part of composing existing operators to create a new operator is that one no longer need to write the backward function. It will be also much easier to help understand the source code. For example, comparing against the above implementation, the backward function of [BatchNorm](https://github.com/dmlc/mxnet/blob/master/src/operator/batch_norm-inl.h) is quite daunting to read.

**Comment by piiswrong** [comment]
The same syntax works for mxnet too. But this is the reason why TF is slow. We don't recommend it unless you only want a quick hacky a solution

**Comment by taoari** [comment]
@piiswrong Very much thank you! I did not notice that it is a speed reason.

**Comment by taoari** [comment]
@piiswrong Any example to do this in the C++ source codes to get around the speed issue?

**Comment by formath** [comment]
I think it don't support that because the inner struct or class in c++ can't support composed operator. Maybe you should add a new operator. Then copy and merge the existing operators's forward and backward code to the new one.

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
